### PR TITLE
[SPARK-48586][SS] Remove lock acquisition in doMaintenance() by making a deep copy of file mappings in RocksDBFileManager in load()

### PR DESCRIPTION
**Pull request by riyaverm-db**
### What changes were proposed in this pull request?
When change log checkpointing is enabled, the lock of the **RocksDB** state store is acquired when uploading the snapshot inside maintenance tasks, which causes lock contention between query processing tasks and state maintenance thread. This PR fixes lock contention issue introduced by https://github.com/apache/spark/pull/45724. 

The changes include:
1. Removing lock acquisition in `doMaintenance()`
2. Adding a `copyFileMappings()` method to **RocksDBFileManager**,  and using this method to deep copy the file manager state, specifically the file mappings `versionToRocksDBFiles` and `localFilesToDfsFiles`,  in `load()`
3. Capture the reference to the file mappings  in `commit()`.

 
### Why are the changes needed?
We want to eliminate lock contention to decrease latency of streaming queries so lock acquisition inside maintenance tasks should be avoided. This can introduce race conditions between task and maintenance threads. By making a deep copy of `versionToRocksDBFiles` and `localFilesToDfsFiles` in **RocksDBFileManager**, we can ensure that the file manager state is not updated by task thread when background snapshot uploading tasks attempt to upload a snapshot.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?

Added unit test cases.

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by riyaverm-db** [comment]
@chaoqin-li1123

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala [hunk 1]
```diff
@@ -874,6 +875,48 @@ class RocksDBSuite extends AlsoTestWithChangelogCheckpointingEnabled with Shared
...
+    val dfsRootDir = new File(Utils.createTempDir().getAbsolutePath + "/state/1/1")
+    val originalFileManager = new RocksDBFileManager(
+      dfsRootDir.getAbsolutePath, Utils.createTempDir(), new Configuration)
+
+    val file1 = RocksDBSstFile("001.sst", "dfs1", 100L)
+    val file2 = RocksDBSstFile("002.sst", "dfs2", 100L)
+
+    // Get access to the private ConcurrentHashMap
+    val origVersionToRocksDBFilesMap: ConcurrentHashMap[Long, Seq[RocksDBImmutableFile]]
+    = getPrivateField(originalFileManager, "versionToRocksDBFiles")
```
Comment (chaoqin-li1123) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala:888]: Try to verify the independence without accessing private field.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala [hunk 2]
```diff
@@ -135,7 +135,6 @@ class RocksDBFileManager(
 
   private val versionToRocksDBFiles = new ConcurrentHashMap[Long, Seq[RocksDBImmutableFile]]
 
-
```
Comment (chaoqin-li1123) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:138]: Unnecessary change.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala [hunk 3]
```diff
@@ -181,6 +190,7 @@ class RocksDB(
     try {
       if (loadedVersion != version) {
         closeDB()
+        fileManager = fileManager.deepCopy()
```
Comment (chaoqin-li1123) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:193]: Can we comment here why we want need to make a deep copy here?

Comment (riyaverm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:195]: Added comment

**Comment by riyaverm-db** [comment]
@sahnib

**Comment by chaoqin-li1123** [comment]
Can you fill in testing description in the PR description?
Remove the WIP from title and finalize the title and description when it is ready for review.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala [hunk 4]
```diff
@@ -874,6 +877,74 @@ class RocksDBSuite extends AlsoTestWithChangelogCheckpointingEnabled with Shared
...
+      assert(originalFileManager.latestSaveCheckpointMetrics.filesCopied == 0L)
+      assert(originalFileManager.latestSaveCheckpointMetrics.bytesCopied == 0L)
+
+      // Copied file manager should not skip since these are new files
+      assert(copiedFileManager.latestSaveCheckpointMetrics.filesCopied == 3L)
+      assert(copiedFileManager.latestSaveCheckpointMetrics.bytesCopied == 30L)
+    }
+  }
+
+  testWithChangelogCheckpointingEnabled("RocksDBFileManager: eliminate lock contention") {
```
Comment (chaoqin-li1123) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala:915]: Can we be more specific and name it "background snapshot upload doesn't acquire rocksdb instance lock"?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala [hunk 5]
```diff
@@ -874,6 +877,74 @@ class RocksDBSuite extends AlsoTestWithChangelogCheckpointingEnabled with Shared
...
+      // Run doMaintenance in another thread
+      val maintenanceFuture = Future {
+        db.doMaintenance()
+      }
+
+      val timeout = 5.seconds
+
+      // Ensure that maintenance task runs without being blocked by task thread
+      ThreadUtils.awaitResult(maintenanceFuture, timeout)
+
```
Comment (chaoqin-li1123) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala:942]: Can we check the checkpoint directory content here and verify snapshot is indeed uploaded?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala [hunk 6]
```diff
@@ -584,14 +595,19 @@ class RocksDB(
...
           cp.createCheckpoint(checkpointDir.toString)
-          // if changelog checkpointing is disabled, the snapshot is uploaded synchronously
-          // inside the uploadSnapshot() called below.
-          // If changelog checkpointing is enabled, snapshot will be uploaded asynchronously
-          // during state store maintenance.
-          latestSnapshot.foreach(_.close())
-          latestSnapshot = Some(
-            RocksDBSnapshot(checkpointDir, newVersion, numKeysOnWritingVersion))
-          lastSnapshotVersion = newVersion
+          synchronized {
```
Comment (sahnib) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:598]: Are we trying to synchronize updates between task thread/maintenance thread here? If so, we need a lock - just synchronized would not work. This code can only be executed by one thread anyways (task thread that holds the acquire lock - so synchronized does not do anything in my opinion).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala [hunk 7]
```diff
@@ -649,11 +665,11 @@ class RocksDB(
       checkpoint
     }
     localCheckpoint match {
-      case Some(RocksDBSnapshot(localDir, version, numKeys)) =>
+      case Some(RocksDBSnapshot(localDir, version, numKeys, previousFileManager)) =>
```
Comment (sahnib) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:668]: [nit] Probably better to name it as `snapshotFileManager`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala (hunk 6 above)
Comment (chaoqin-li1123) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:598]: What is the race condition you are thinking about? Is calling close() on a snapshot being uploaded? In that case we can defer the close to the maintenance thread.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala [hunk 8]
```diff
@@ -649,17 +660,23 @@ class RocksDB(
...
             fileManagerMetrics = fileManager.latestSaveCheckpointMetrics
           }
           logInfo(log"${MDC(LogKeys.LOG_ID, loggingId)}: Upload snapshot of version " +
             log"${MDC(LogKeys.VERSION_NUM, version)}," +
             log" time taken: ${MDC(LogKeys.TIME_UNITS, uploadTime)} ms")
         } finally {
           localCheckpoint.foreach(_.close())
+
+          // Clean up old latestSnapshots
+          while (oldSnapshots.nonEmpty) {
```
Comment (riyaverm-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:676]: Race condition for oldSnapshots here between maintenance operations and commit `addOne()`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala [hunk 9]
```diff
@@ -157,6 +147,29 @@ class RocksDBFileManager(
...
+    val newVersionToRocksDBFiles = new ConcurrentHashMap[Long, Seq[RocksDBImmutableFile]]
+    val newLocalFilesToDfsFiles = new ConcurrentHashMap[String, RocksDBImmutableFile]
+
+    newVersionToRocksDBFiles.putAll(fileMappings.versionToRocksDBFiles)
+    newLocalFilesToDfsFiles.putAll(fileMappings.localFilesToDfsFiles)
+
+    fileMappings = RocksDBFileMappings(newVersionToRocksDBFiles, newLocalFilesToDfsFiles)
+  }
+
+  def captureFileMapReference(): RocksDBFileMappings = {
```
Comment (siying) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:170]: I didn't quite get the purpose of this function.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala [hunk 10]
```diff
@@ -74,14 +74,20 @@ class RocksDB(
     loggingId: String = "",
     useColumnFamilies: Boolean = false) extends Logging {
 
-  case class RocksDBSnapshot(checkpointDir: File, version: Long, numKeys: Long) {
+  case class RocksDBSnapshot(
+    checkpointDir: File,
+    version: Long,
+    numKeys: Long,
+    fileMappings: RocksDBFileMappings) {
```
Comment (siying) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:81]: if we call it `capturedFileMappings` everywhere else, why not call it this name here too?

**Review by siying — APPROVED** [review]
Thanks for working on it. I don't have any other concern anymore.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala (hunk 9 above)
Comment (riyaverm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:170]: It functions as a get method for the RocksDBFileMapping private var. I named it captureFileMapReference for code readability in RocksDB where it is used.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala (hunk 10 above)
Comment (riyaverm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:81]: Must have missed this, will update.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala [hunk 11]
```diff
@@ -588,9 +596,12 @@ class RocksDB(
...
           // If changelog checkpointing is enabled, snapshot will be uploaded asynchronously
           // during state store maintenance.
-          latestSnapshot.foreach(_.close())
+          oldSnapshots += latestSnapshot
           latestSnapshot = Some(
-            RocksDBSnapshot(checkpointDir, newVersion, numKeysOnWritingVersion))
+            RocksDBSnapshot(checkpointDir,
+              newVersion,
+              numKeysOnWritingVersion,
+              fileManager.captureFileMapReference()))
```
Comment (sahnib) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:604]: I think we need to synchronize this part to prevent race with maintenance thread on mutating latestSnapshot.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala [hunk 12]
```diff
@@ -588,9 +596,12 @@ class RocksDB(
           // inside the uploadSnapshot() called below.
           // If changelog checkpointing is enabled, snapshot will be uploaded asynchronously
           // during state store maintenance.
-          latestSnapshot.foreach(_.close())
+          oldSnapshots += latestSnapshot
```
Comment (sahnib) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:599]: Do we need to check if `latestSnapshot` is None before we append it to the list?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala [hunk 13]
```diff
@@ -74,14 +74,20 @@ class RocksDB(
...
+    checkpointDir: File,
+    version: Long,
+    numKeys: Long,
+    capturedFileMappings: RocksDBFileMappings) {
     def close(): Unit = {
       silentDeleteRecursively(checkpointDir, s"Free up local checkpoint of snapshot $version")
     }
   }
 
+
```
Comment (sahnib) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:87]: [nit] extra line.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala (hunk 6 above)
Comment (sahnib) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:599]: Sorry I missed your comment  @chaoqin-li1123. The race condition is basically assigning to latestSnapshot variable. We assign the value of variable to newly created snapshot here, and set the variable to None in maintenance thread. Both these need to be synchronized, I think. 

It seems to have been fixed in the latest revision.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala [hunk 14]
```diff
@@ -643,23 +658,36 @@ class RocksDB(
   }
 
   private def uploadSnapshot(): Unit = {
+    var oldSnapshotsImmutable: List[RocksDBSnapshot] = Nil
     val localCheckpoint = synchronized {
       val checkpoint = latestSnapshot
       latestSnapshot = None
```
Comment (sahnib) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:664]: [unrelated to this PR - just a note here]. I think if we fail to write to DFS (lets say due to some timeout/network error), we lose the snapshot as we assign the variable to `None` here. Probably better to assign this variable to `None` after a successful upload. 

However, I think its inherently complex to do what I mentioned in above paragraph, because the task thread can change the latestSnapshot to a new snapshot while upload is happening. I guess maybe its okay to skip a snapshot on transient error (for simplicity) and upload the next snapshot. 

Curious, if you have any further thoughts on this @riyaverm-db @chaoqin-li1123 .

**Review by sahnib — APPROVED** [review]
LGTM. thanks for making these changes. Its super helpful to avoid contention between maintenance and task thread for performance, and ensuring snapshots keep on uploading at regular cadence.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala (hunk 14 above)
Comment (chaoqin-li1123) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:663]: Snapshot uploading is best effort attempt anyway. It should be fine we we skip some uploading due to transient error.

**Review by chaoqin-li1123 — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala [hunk 15]
```diff
@@ -74,14 +74,19 @@ class RocksDB(
     loggingId: String = "",
     useColumnFamilies: Boolean = false) extends Logging {
 
-  case class RocksDBSnapshot(checkpointDir: File, version: Long, numKeys: Long) {
+  case class RocksDBSnapshot(
+    checkpointDir: File,
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:78]: nit: 2 more spaces (4 spaces for params in multi-lines)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala [hunk 16]
```diff
@@ -204,11 +217,15 @@ class RocksDBFileManager(
   def latestSaveCheckpointMetrics: RocksDBFileManagerMetrics = saveCheckpointMetrics
 
   /** Save all the files in given local checkpoint directory as a committed version in DFS */
-  def saveCheckpointToDfs(checkpointDir: File, version: Long, numKeys: Long): Unit = {
+  def saveCheckpointToDfs(
+    checkpointDir: File,
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:221]: nit: same here, 2 more spaces

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala [hunk 17]
```diff
@@ -749,6 +768,20 @@ class RocksDBFileManager(
...
+ * @param versionToRocksDBFiles Mapping of RocksDB files used across versions for maintenance
+ * @param localFilesToDfsFiles Mapping of the exact Dfs file used to create a local SST file
+ * The reason localFilesToDfsFiles is a separate map because versionToRocksDBFiles can contain
+ *  multiple similar SST files to a particular local file (for example 1.sst can map to 1-UUID1.sst
+ * in v1 and 1-UUID2.sst in v2). We need to capture the exact file used to ensure Version ID
+ * compatibility across SST files and RocksDB manifest.
+ */
+
+case class RocksDBFileMappings(
+  versionToRocksDBFiles: ConcurrentHashMap[Long, Seq[RocksDBImmutableFile]],
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:782]: nit: 2 more spaces

**Review by HeartSaVioR — APPROVED** [review]
Given there are three approvals and the PR has been up for 2 weeks, I wouldn't require round of review to fix nits. Please file a new FOLLOWUP PR to address the comments.

+1

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master/3.5/3.4.

**Comment by HeartSaVioR** [comment]
@riyaverm-db Looks like there is merge conflict in 3.5 (and probably 3.4 as well). Could you please help crafting a PR for 3.5 and 3.4? Thanks in advance!
